In [184]:
import numpy as np


from bokeh.io import show, output_notebook
from bokeh.plotting import figure
from bokeh.layouts import row
output_notebook(hide_banner=True)

from scipy.optimize import newton

# Méthode de Newton

### Implémentation de la méthode

In [185]:
def my_newton(f, df, x0, xstar, tol=1.e-12, nitmax=50):
   
    # initialisation
    x = np.zeros(nitmax+1)
    x[0] = x0
    
    # iteration de Newton        
    for i in range(1, nitmax+1):
        x[i] = x[i-1] - f(x[i-1])/df(x[i-1])
        #if ( abs(x[i] - xstar) < tol) : break
        if ( abs(f(x[i])) < tol ): break
        if (abs(x[i]-x[i-1]) < tol) : break

    return x[0:i+1]

### Ordre de convergence

Image élément par élément d'un vecteur par une fonction de variable réelle

In [186]:
def elt_par_elt(f,x) :
    y = np.zeros(x.size)
    for i in range(x.size) :
        y[i] = f(x[i])
    return y

Test de la fonction : $f(x) = e^x - 2$

In [187]:
def f(x):
    return np.exp(x) - 2

def df(x):
    return np.exp(x) 

In [190]:
xstar = np.log(2)
xsol = my_newton(f, df, x0=0., xstar=xstar, tol=1.e-12, nitmax=50)

print(f"Convergence de l'algorithme de Newton en {xsol.size-1} itérations")
print(f"--> solution de f(x) = 0 obtenue pour  x = {xsol[-1]}")

err = np.abs(xsol - xstar)
im = elt_par_elt(f,xsol)

fig1 = figure(width=490, height=300, y_axis_type="log")
fig1.x(range(xsol.size), err, size=5)

fig2 = figure(width=490, height=300, x_axis_type="log", y_axis_type="log")
fig2.x(err[:-1], err[1:], size=10) #err(n+1) en fctn de err(n)
fig2.x(im[:-1], im[1:], size=10, color="orange", legend="erreur résiduelle") #f(x_n+1) en fctn de f(x_n)
fig2.line(err[:-1], err[:-1],  color="yellow", legend="ordre 1")
fig2.line(err[:-1], err[:-1]*err[:-1],  color="red", legend="ordre 2")
fig2.line(err[:-1], err[:-1]*err[:-1]*err[:-1],  color="green", legend="ordre 3")

show(row(fig1, fig2))

Convergence de l'algorithme de Newton en 5 itérations
--> solution de f(x) = 0 obtenue pour  x = 0.6931471805600254


Test de la fonction : $g(x) = \displaystyle \frac{x}{\sqrt{1+x^2}}$

In [191]:
def g(x):
    return x/(np.sqrt(1+x*x))

def dg(x):
    return 1/(np.power(1+x*x, 1/3))

In [193]:
xstar = 0.
xsol = my_newton(g, dg, x0=0.5, xstar=xstar, tol=1.e-12, nitmax=50)
print(f"Convergence de l'algorithme de Newton en {xsol.size-1} itérations")
print(f"--> solution de g(x) = 0 obtenue pour  x = {xsol[-1]}")

err = np.abs(xsol - xstar)
im = elt_par_elt(g,xsol)

fig1 = figure(width=490, height=300, y_axis_type="log")
fig1.x(range(xsol.size), err, size=5)

fig2 = figure(width=490, height=300, x_axis_type="log", y_axis_type="log")
fig2.x(err[:-1], err[1:], size=10) #err(n+1) en fctn de err(n)
fig2.x(im[:-1], im[1:], size=5, color="orange", legend="erreur résiduelle") #g(x_n+1) en fctn de g(x_n)

fig2.line(err[:-1], err[:-1],  color="yellow", legend="ordre 1")
fig2.line(err[:-1], err[:-1]*err[:-1],  color="red", legend="ordre 2")
fig2.line(err[:-1], err[:-1]*err[:-1]*err[:-1],  color="green", legend="ordre 3")

show(row(fig1, fig2))

Convergence de l'algorithme de Newton en 3 itérations
--> solution de g(x) = 0 obtenue pour  x = 1.73429995950318e-19


Test de la fonction : $h(x) = x^3$

In [194]:
def h(x):
    return x*x*x

def dh(x):
    return 3*x*x

In [195]:
xstar = 0.
xsol = my_newton(h, dh, x0=0.5, xstar=xstar, tol=1.e-12, nitmax=100)
print(f"Convergence de l'algorithme de Newton en {xsol.size-1} itérations")
print(f"--> solution de h(x) = 0 obtenue pour  x = {xsol[-1]}")

err = np.abs(xsol - xstar)
im = elt_par_elt(h,xsol)

fig1 = figure(width=490, height=300, y_axis_type="log")
fig1.x(range(xsol.size), err, size=5)

fig2 = figure(width=490, height=300, x_axis_type="log", y_axis_type="log")
fig2.x(err[:-1], err[1:], size=10) #err(n+1) en fctn de err(n)
fig2.x(im[:-1], im[1:], size=10, color="orange",legend="erreur résiduelle") #h(x_n+1) en fctn de h(x_n)

fig2.line(err[:-1], err[:-1],  color="yellow", legend="ordre 1")
fig2.line(err[:-1], err[:-1]*err[:-1],  color="red", legend="ordre 2")
fig2.line(err[:-1], err[:-1]*err[:-1]*err[:-1],  color="green", legend="ordre 3")



show(row(fig1, fig2))

Convergence de l'algorithme de Newton en 22 itérations
--> solution de h(x) = 0 obtenue pour  x = 6.682859107149278e-05


**Commenter ces résulats et observer également l'erreur résiduelle.** 

J'ai défini au début de cette partie une fonction elt_par_elt qui prend en argument une fonction $f$ à variable réelle et un vecteur $X$, et renvoie le vecteur des images des éléments de $X$ par $f$. (pour observer l'erreur résiduelle)

Ensuite j'ai représenté sur la figure en échelle log-log chacune des droites correspondant aux ordres 1, 2 et 3, ainsi que l'érreur résiduelle (croix oranges).

Cette échelle nous permet bien de vérifier l'ordre de la méthode utilisée car on a
$\epsilon_{n+1} \leq C\epsilon_n^{\mu}$, et par conséquent $log(\epsilon_{n+1}) \leq log(C)+\mu .log(\epsilon_n)$.
L'ordre devient alors la pente de la courbe.

Les courbes obtenues permettent d'affirmer les résultats théoriques : les ordres de $f$, $g$ et $h$ sont respéctivement 2, 3 et 1. Et on remarque de plus que pour $f$ et $g$, ayant des dérivées non nulles en leur zéros, l'erreur résiduelle a le même ordre de convergence que $\epsilon_n$.

**Comparer ces résultats avec la fonction newton de Scipy (https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.newton.html)**

In [196]:
xsol = newton(f, x0=0, fprime=df, args=(), tol=1.e-12, maxiter=50, fprime2=None, x1=None, rtol=0.0, full_output=True, disp=True)
print( "pour la fonction f : ",xsol,"\n")

xsol = newton(g, x0=0.5, fprime=dg, args=(), tol=1.e-12, maxiter=50, fprime2=None, x1=None, rtol=0.0, full_output=True, disp=True)
print( "pour la fonction f : ",xsol,"\n")

xsol = newton(h, x0=0.5, fprime=dh, args=(), tol=1.48e-08, maxiter=50, fprime2=None, x1=None, rtol=0.0, full_output=True, disp=True)
print( "pour la fonction f : ",xsol,"\n")

pour la fonction f :  (0.6931471805599453,       converged: True
           flag: 'converged'
 function_calls: 12
     iterations: 6
           root: 0.6931471805599453) 

pour la fonction f :  (0.0,       converged: True
           flag: 'converged'
 function_calls: 8
     iterations: 4
           root: 0.0) 

pour la fonction f :  (2.0097272630703606e-08,       converged: True
           flag: 'converged'
 function_calls: 84
     iterations: 42
           root: 2.0097272630703606e-08) 



Les nombres d'itérations de my_newton() et de numpy.scipy.newton() sont du même ordre de grandeur, ce qui montre que les ordres de convergence obtenus par les deux fonctions pour chacune des fonctions $f$, $g$ et $h$ sont les mêmes.

### Cas dégénérés

**Implémenter l'algorithme de Newton modifié sous la forme $x_{n+1} = x_n-m\frac{f(x_n)}{f'(x_n)} $**

In [197]:
def my_newton_modif(f, df, m, x0, xstar, tol=1.e-12, nitmax=50):
   
    # initialisation
    x = np.zeros(nitmax+1)
    x[0] = x0
    
    # iteration du Newton modifié       
    for i in range(1, nitmax+1):
        x[i] = x[i-1] - m*f(x[i-1])/df(x[i-1])
        if ( abs(x[i] - xstar) < tol) : break
        if ( abs(f(x[i])) < tol ): break
        if (abs(x[i]-x[i-1]) < tol) : break

    return x[0:i+1]

**Vérifier numériquement le taux de convergence de la méthode de Newton modifié pour la racine de la fonction : $k(x) = x^2(x^2+2)$. Commenter**

In [198]:
def k(x):
    return (x*x)*(x*x+2)

def dk(x):
    return 4*x*x*x + 4*x

In [200]:
xstar = 0.
xsol = my_newton_modif(k, dk, 2, x0=0.5, xstar=xstar, tol=1.e-15, nitmax=50)
print(f"Convergence de l'algorithme de Newton en {xsol.size-1} itérations")
print(f"--> solution de k(x) = 0 obtenue pour  x = {xsol[-1]}")

err = np.abs(xsol - xstar)
im = elt_par_elt(k,xsol)

fig1 = figure(width=490, height=300, y_axis_type="log")
fig1.x(range(xsol.size), err, size=5)

fig2 = figure(width=490, height=300, x_axis_type="log", y_axis_type="log")
fig2.x(err[:-1], err[1:], size=10) #err(n+1) en fctn de err(n)
fig2.x(im[:-1], im[1:], size=5, color="orange",legend="erreur résiduelle") #g(x_n+1) en fctn de g(x_n)

fig2.line(err[:-1], err[:-1],  color="yellow", legend="ordre 1")
fig2.line(err[:-1], err[:-1]*err[:-1],  color="red", legend="ordre 2")
fig2.line(err[:-1], err[:-1]*err[:-1]*err[:-1],  color="green", legend="ordre 3")

show(row(fig1, fig2))

Convergence de l'algorithme de Newton en 3 itérations
--> solution de k(x) = 0 obtenue pour  x = 1.2115933527723916e-13


La convergence obtenue est d'ordre 3 ! et donc mieux même qu'une convergence quadratique, ce qui affirme le résultat théorique.

# Méthode de la sécante

### Implémentation de la méthode

In [209]:
def taux_accroissement(f,x,y) :
    return (f(x)-f(y))/(x-y)

def my_secante(f, x0, x1, xstar, tol=1.e-12, nitmax=50):
   
    # initialisation
    x = np.zeros(nitmax+2)
    x[0] = x0
    x[1] = x1
    
    # itération de la methode de la secante
    for i in range(2,nitmax+2) :
        x[i] = x[i-1] - f(x[i-1])/taux_accroissement(f,x[i-1],x[i-2])
        if abs(x[i]-xstar)<tol : break
    
    return x[0:i+1]

**Vérifier si la méthode de la sécante a un ordre de convergence d'au moins $\rho  = \displaystyle \frac{\sqrt{5}+1}{2}$ pour les fonctions $f(x)$, $g(x)$ et $h(x)$. Commenter.**

Test de la fonction $f$ :

In [211]:
rho = (1+5**0.5)/2

xstar = np.log(2)
xsol = my_secante(f, x0=0.6,x1=0.7, xstar=xstar, tol=1.e-12, nitmax=50)
print(f"Convergence de la méthode de la sécante en {xsol.size-1} itérations")
print(f"--> solution de f(x) = 0 obtenue pour  x = {xsol[-1]}")

err = np.abs(xsol - xstar)

fig2 = figure(width=490, height=300, x_axis_type="log", y_axis_type="log")
fig2.x(err[:-1], err[1:], size=5)
fig2.line(err[:-1], err[:-1]**rho,  color="green", legend="ordre rho")

show(fig2)

Convergence de la méthode de la sécante en 5 itérations
--> solution de f(x) = 0 obtenue pour  x = 0.6931471805599452


Test de la fonction $g$ :

In [212]:
xstar = 0
xsol = my_secante(g, x0=0.3,x1=0.4, xstar=xstar, tol=1.e-12, nitmax=50)
print(f"Convergence de la méthode de la sécante en {xsol.size-1} itérations")
print(f"--> solution de g(x) = 0 obtenue pour  x = {xsol[-1]}")

err = np.abs(xsol - xstar)


fig2 = figure(width=490, height=300, x_axis_type="log", y_axis_type="log")
fig2.x(err[:-1], err[1:], size=5)
fig2.line(err[:-1], err[:-1]**rho,  color="green", legend="ordre rho")

show(fig2)

Convergence de la méthode de la sécante en 6 itérations
--> solution de g(x) = 0 obtenue pour  x = -2.638092230933279e-23


Test pour la fonction $h$ :

In [213]:
xstar = 0
xsol = my_secante(h, x0=0.3,x1=0.4, xstar=xstar, tol=1.e-12, nitmax=50)
print(f"Convergence de la méthode de la sécante en {xsol.size-1} itérations")
print(f"--> solution de h(x) = 0 obtenue pour  x = {xsol[-1]}")

err = np.abs(xsol - xstar)

fig1 = figure(width=490, height=300, x_axis_type="log", y_axis_type="log")
fig1.x(err[:-1], err[1:], size=5)
fig1.line(err[:-1], err[:-1],  color="yellow", legend="ordre 1")
fig1.line(err[:-1], err[:-1]**rho,  color="green", legend="ordre rho")

show(fig1)

Convergence de la méthode de la sécante en 51 itérations
--> solution de h(x) = 0 obtenue pour  x = 2.517346967800027e-07


Nous obtenons effectivement un ordre d'au moins $\rho$ pour $g$ et $f$ car $f(0) = 0$, $f'(0) \ne 0$ et $g(0) = 0$, $g'(0) \ne 0$.

Par contre, nous obtenons un ordre linéaire pour $h$, ce qui ne contredit pas la théorie car $h(0)=0$ et $h'(0)=0$.

# Bissection-Trissection-hybride

### Implémentation de la méthode

In [214]:
def my_bissection(f, a, b, xstar=None, tol=1.e-12, nitmax=50):
   
    # initialisation
    x = np.zeros(nitmax+1)
    x[0] = (a+b)/2
    
    # iteration de la méthode de bissection
    for i in range(1,nitmax+1) :
        if f(x[i-1])*f(a)<=0 :
            b = x[i-1]
        else :
            a = x[i-1]
        x[i] = (a+b)/2
        if b-a<tol : break
        if abs(x[i]-xstar)<tol : break
        if abs(f(x[i]))<tol : break

    return x[0:i+1]

**Tester la méthode avec la fonction $m(x) = ln(x)+2$. Commenter.**

In [215]:
def m(x) :
    return np.log(x)+2

def dm(x) :
    return 1/x

In [218]:
xstar = np.exp(-2)

## Methode de la bissection
xsol = my_bissection(m, 0.01, 1, xstar, tol=1.e-12, nitmax=50)
print(f"Convergence de la méthode de la bissection en {xsol.size-1} itérations")
print(f"--> solution de m(x) = 0 obtenue pour  x = {xsol[-1]}\n")
err = np.abs(xsol - xstar)

## Methode de Newton
xsol_newton = my_newton(m, dm, x0=0.01, xstar=xstar, tol=1.e-10, nitmax=100)
print(f"Convergence de l'algorithme de Newton en {xsol_newton.size-1} itérations")
print(f"--> solution de m(x) = 0 obtenue pour  x = {xsol_newton[-1]}\n")
err_newton = np.abs(xsol_newton - xstar)

err_sol = abs(xsol[-1]-xsol_newton[-1])
print ("l'erreur entre les solutions trouvées par deux méthodes est : ",err_sol)

fig1 = figure(width=490, height=300, x_axis_type="log", y_axis_type="log")
fig1.x(err[:-1], err[1:], size=5, legend="Bissection")
fig1.x(err_newton[:-1], err_newton[1:],color="orange", size=5, legend="Newton")
fig1.line(err[:-1], err[:-1],  color="yellow", legend="ordre 1")
fig1.line(err[:-1], err[:-1]*err[:-1],  color="red", legend="ordre 2")

show(fig1)

Convergence de la méthode de la bissection en 38 itérations
--> solution de m(x) = 0 obtenue pour  x = 0.13533528323672728

Convergence de l'algorithme de Newton en 6 itérations
--> solution de m(x) = 0 obtenue pour  x = 0.1353352832336045

l'erreur entre les solutions trouvées par deux méthodes est :  3.122779812514409e-12


L'erreur entre les solutions obtenues par chacune des deux méthode est de l'ordre de $10^{-12}$, on aboutit donc au même bon résultat. Toutefois, avec les conditions initiales que j'ai choisies, la méthode de Newton converge en 6 itérations alors que la méthode dichotomique converge en 38 itérations.

En effet, La méthode de Newton est d'ordre 2, et la méthode de bissection est d'ordre 1.

# algorithme hybride Bissection-Newton

**Idée de l'algorithme hybride :** Cet algorithme consiste à appliquer la méthode de Newton tout en garantissant la convergence, en imposant qu'on s'approche de la racine $x_*$ après chaque itération grâce à une méthode dichotomique.

En terme de qualité, cet algorithme convergence au moins au même ordre de convergence que l'algorithme de Newton, et donc pour une fonction de classe $C^2$ ayant un zéro non dégénéré, la convergence est au moins quadratique.

In [220]:
def my_hybride(f, df, a, b, x0, tol=1.e-12, nitmax=50) :
    
    # initialisation
    x = np.zeros(nitmax+1)
    x[0] = x0
    
    #itération
    for i in range(1,nitmax+1) :
        y = x[i-1]-f(x[i-1])/df(x[i-1])
        # construction de x[i]
        if a<=y and y<=b : x[i]=y
        else : x[i]=(a+b)/2
        # modification de a et b
        if f(a)*f(x[i])<=0 : b=x[i]
        else : a=x[i]
        
        if ( abs(f(x[i])) < tol ): break
        if (abs(x[i]-x[i-1]) < tol) : break

    return x[0:i+1]

Considérer la fonction $f(x) = x^5-x+1$ sur l'intervalle $[-2.5,2.5]$.
* Tester la méthode de Newton pour cette fonction avec $x_0=-1$ et $x_0=1$. En cas de non convergence, expliquer le comportement observé.
* Tester votre algorithme hybride Bissection-Newton sur la fonction $f$. Représenter graphiquement la courbe de convergence et commenter le comportement de votre algorithme.

In [221]:
def f(x) :
    return x**5-x+1

def df(x) :
    return 5*x**4-1

In [222]:
# Pour x0 = 1 :
xsol = my_newton(f, df, x0=1, xstar=None, tol=1.e-12, nitmax=100)
print(f"Pour x0 = 1 : convergence de l'algorithme de Newton en {xsol.size-1} itérations")
print(f"--> solution de f(x) = 0 obtenue pour  x* = {xsol[-1]}")
print("--> Erreur résiduelle : f(x*) = ",f(xsol[-1]))

# Pour x0 = -1 :
xsol2 = my_newton(f, df, x0=-1, xstar=None, tol=1.e-12, nitmax=50)
print(f"Pour x0 = -1 : convergence de l'algorithme de Newton en {xsol2.size-1} itérations")
print(f"--> solution de f(x) = 0 obtenue pour  x* = {xsol2[-1]}")
print("--> Erreur résiduelle : f(x*) = ",f(xsol2[-1]))

Pour x0 = 1 : convergence de l'algorithme de Newton en 100 itérations
--> solution de f(x) = 0 obtenue pour  x* = 0.7503218281592572
--> Erreur résiduelle : f(x*) =  0.4874924386834848
Pour x0 = -1 : convergence de l'algorithme de Newton en 5 itérations
--> solution de f(x) = 0 obtenue pour  x* = -1.1673039782614396
--> Erreur résiduelle : f(x*) =  -1.7341683644644945e-13


* Pour $x_0 = -1$, on a convergence de l'algorithme, et le résultat obtenu est satisfaisant car l'erreur résiduelle est suffisamment petite.
* Pour $x_0 = 1$, l'algorithme ne converge pas.

Ce comportement est dû à la distance entre le point initial et le zéro de la fonction : -1 est assez proche de $x_* \simeq -1.167$, mais 1 ne l'est pas.

**Test de l'algorithme hybride Bissection-Newton pour la fonction $f$**

In [163]:
xsol = my_hybride(f, df, a=-2.5, b=2.5, x0=1, tol=1.e-12, nitmax=100)

print(f"Pour x0 = 1 : convergence de l'algorithme de Newton en {xsol.size-1} itérations")
print(f"--> solution de f(x) = 0 obtenue pour  x* = {xsol[-1]}")
print("--> Erreur résiduelle : f(x*) = ",f(xsol[-1]))

xstar = xsol[-1]
err = np.abs(xsol - xstar)


fig2 = figure(width=490, height=300, x_axis_type="log", y_axis_type="log")
fig2.x(err[:-1], err[1:], size=10) #err(n+1) en fctn de err(n)
fig2.line(err[:-1], err[:-1],  color="yellow", legend="ordre 1")
fig2.line(err[:-1], err[:-1]*err[:-1],  color="red", legend="ordre 2")
fig2.line(err[:-1], err[:-1]*err[:-1]*err[:-1],  color="green", legend="ordre 3")

show(fig2)

Pour x0 = 1 : convergence de l'algorithme de Newton en 8 itérations
--> solution de f(x) = 0 obtenue pour  x* = -1.1673039782614187
--> Erreur résiduelle : f(x*) =  -6.661338147750939e-16


On remarque que l'algorithme hybride bissection-Newton résout bien le problème de convergence rencontré avec l'algorithme de Newton : l'algorithme converge quelque soit la condition initiale $x_0$.

En plus, sa convergence est d'ordre **quadratique**.